stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -0,0 +1,954 @@
+
+def intName(name, signed):
+    return ('' if signed else 'U') + baseIntName(name)
+
+floatName = { 32:'Float', 64:'Double', 80:'Float80' }
+
+def cFuncSuffix(bits):
+    if bits == 32:
+        return 'f'
+    if bits == 64:
+        return ''
+    if bits == 80:
+        return 'l'
+        
+def intFormatFix(bits):
+    if bits == 'Int':
+        return int(CMAKE_SIZEOF_VOID_P) * 8
+    return bits
+}%
+
+//  TODO: remove once integer proposal is available ----------------------------
+%for bits in [32,64]:
+extension UInt${bits} {
+  var signBitIndex: Int {
+    return ${bits-1} - Int(Int${bits}(Builtin.int_ctlz_Int${bits}(self._value, false._value)))
+  }
+  var countTrailingZeros: Int {
+    return Int(Int${bits}(Builtin.int_cttz_Int${bits}(self._value, false._value)))
+  }
+}
+%end
+
+% for bits in allFloatBits:
+%{
+Self = floatName[bits]
+SignificandSize = {32:32, 64:64, 80:64}[bits]
+RawSignificand = 'UInt' + str(SignificandSize)
+
+if Self == 'Float':
+    SelfDocComment = '''\
+/// A single-precision floating-point value type.'''
+
+elif Self == 'Double':
+    SelfDocComment = '''\
+/// A double-precision floating-point value type.'''
+
+elif Self == 'Float80':
+    SelfDocComment = '''\
+/// An extended-precision floating-point value type.'''
+
+else:
+    raise ValueError('Unhandled float type.')
+}%
+
+% if bits == 80:
+#if !os(Windows) && (arch(i386) || arch(x86_64))
+% end
+
+${SelfDocComment}
+public struct ${Self} {
+  public // @testable
+  var _value: Builtin.FPIEEE${bits}
+
+  /// Create an instance initialized to zero.
+  @_transparent public
+  init() {
+    let zero: Int64 = 0
+    self._value = Builtin.sitofp_Int64_FPIEEE${bits}(zero._value)
+  }
+
+  @_transparent 
+  public // @testable
+  init(_bits v: Builtin.FPIEEE${bits}) {
+    self._value = v
+  }
+
+  /// Create an instance initialized to `value`.
+  @_transparent public
+  init(_ value: ${Self}) { self = value }
+}
+
+extension ${Self} : CustomStringConvertible {
+  /// A textual representation of `self`.
+  public var description: String {
+    return _float${bits}ToString(self, debug: false)
+  }
+}
+
+extension ${Self} : CustomDebugStringConvertible {
+  /// A textual representation of `self`.
+  public var debugDescription: String {
+    return _float${bits}ToString(self, debug: true)
+  }
+}
+
+extension ${Self}: BinaryFloatingPoint {
+
+  public typealias Exponent = Int
+  public typealias RawSignificand = ${RawSignificand}
+
+  public static var exponentBitCount: Int {
+    return ${{32:8, 64:11, 80:15}[bits]}
+  }
+
+  public static var significandBitCount: Int {
+    return ${{32:23, 64:52, 80:63}[bits]}
+  }
+
+  //  Implementation details.
+  static let _infinityExponent = 1 << UInt(exponentBitCount) - 1
+  static let _exponentBias = _infinityExponent >> 1
+  static let _significandMask = 1 << ${RawSignificand}(significandBitCount) - 1
+  static let _quietNaNMask = 1 << ${RawSignificand}(significandBitCount - 1)
```

Comment:
Swift does not have a notion of a constant expression, so these lets are lazily initialized.  The optimizer inlines the expressions in the optimized build, but we are also running tests with a debug library (to make sure we can debug the library), where accessing these lets would involve a `dispatch_once` or an equivalent.

The best way to write them is to use a getter, possibly always-inline:

```swift
static var foo: Int { @inline(__always) get { return ... } }
```

This will be fast in release and debug builds.